docs/design/sql/03-network-protocol.md

```diff
@@ -0,0 +1,402 @@
+discusses the existing networking infrastructure of the Hazelcast cluster and whether it satisfies our design
+principles. Section 3 explains the protocol.
+
+## 1 Design Considerations
+
+In this section, we summarize the fundamental principles that influence the protocol design:
+1. Low latency
+1. Fail-fast
+1. Ordered data exchange
+
+### 1.1 Low Latency
+
+Hazelcast Mustang is a modern distributed SQL engine that targets the OLTP workloads for in-memory data. We expect
+that the engine will be used mostly for relatively short queries, taking milliseconds or seconds to complete. Because network
+latency is still relatively as of 2020, the protocol should be designed to minimize the number of blocking network calls,
+when one member waits for the other.
+
+### 1.2 Fail-Fast
+
+Distributed systems must account for network and hardware failures. Query execution is a stateful process. Whenever a participant
+of a distributed query goes down, there are two options: try tolerating the failure and continue execution, or fail it.
+
+The system needs to track the progress of query execution to tolerate the failure of a query participant. An example of such a
+system is Google Spanner [[1]], which uses restart tokens to track the progress of the query, hiding network failures from
+users. Another approach is to persist intermediate state to disk, which is used in OLAP engines where query restart is not an
+appropriate choice.
+
+While the idea of hiding intermittent failures from users is compelling from the usability point of view, it comes at a massive
+engineering cost, as mentioned by Google Spanner engineers. The Hazelcast Mustang engine is designed with the **fail-fast**
+behavior in mind: whenever a failure occurs, the query is terminated, and a proper transient error is delivered to the user.
+The user should restart the query manually then. The query will fail in the following scenarios:
+1. Member executing the query (participant) has left the cluster
+1. Temporal network problem between two members which resulted in a broken TCP connection
+
+We understand that fail-fast behavior might not be the best solution for some use cases, including machine shutdowns in cloud
+environments. We, therefore, do not reject the idea of implementing transparent handling of network errors in future releases.
+Instead, we treat the fail-fast behavior as a starting point for future improvements, which provides a good trade-off between
+the added value and implementation complexity.
+
+### 1.3 Ordered Data Exchange
+
+The Hazelcast Mustang engine employs the **Volcano Model** approach, as described in [[2]]. Data flows from children relational
+operators to parents operators. The order of rows transferred between operators is essential for correctness. For example, a
+parent operator may expect the child operator to produce rows that are sorted on a particular attribute.
```

Comment:
Is the order enforced for all operators or only for operators really requiring it?

Reply:
The order is enforced for all operators because some optimizer decisions may rely on it. E.g. the optimizer may decide to eliminate the sorting for the `sort <- project <- join <- filter <- index scan` pipeline, so every operator should preserve the order.

In principle, it is possible to relax the order for operators without collation, but that would require adding collation metadata to all operators. The benefit is questionable since all current connection implementations provide ordering anyway.